File: algorithms/2_medium/C/00523_continuous_subarray_sum/hashmap.cpp
class Solution {
public:
    static bool checkSubarraySum(const vector<int>& nums, const int k) {
        assert(k >= 1 && std::ssize(nums) <= 100'000);
        const int32_t len = static_cast<int32_t>(std::ssize(nums));
        auto hashmap = unordered_map<int32_t, int32_t>();
        int32_t prefix = 0;

        hashmap[0] = -1;

        for (int32_t i = 0; i < len; i += 1) {
            prefix += nums[i];
            const int32_t pref_mod_k = prefix % k;
            const auto it = hashmap.find(pref_mod_k);

            if (it != hashmap.end()) {
                if (i - it->second >= 2) {
                    return true;
                }
            } else {
                hashmap[pref_mod_k] = i;
            }
        }

        return false;
    }
};

File: algorithms/2_medium/C/00523_continuous_subarray_sum/hashmap.rs
use std::collections::HashMap;

impl Solution {
    pub fn check_subarray_sum(nums: Vec<i32>, k: i32) -> bool {
        assert!(k >= 1 && nums.len() <= 100_000);
        let mut hashmap: HashMap<i32, i32> = HashMap::with_capacity(nums.len() * 12 / 10);
        let mut prefix = 0_i32;

        hashmap.insert(0, -1);

        for (i, num) in nums.into_iter().enumerate() {
            prefix += num;
            let pref_mod_k = prefix % k;

            if let Some(first_seen) = hashmap.get(&pref_mod_k) {
                if (i as i32) - first_seen >= 2 {
                    return true;
                }
            } else {
                hashmap.insert(pref_mod_k, i as i32);
            }
        }

        false
    }
}
